### PR TITLE
:bug: fix #1030 The Spring Boot jar fails to run when using @AiServic…

### PR DESCRIPTION
**Pull request by clear2x**
The Spring Boot jar fails to run when using @AiService due to Reflections[0.10.2]

Relevant Links:   
- https://github.com/langchain4j/langchain4j/issues/1030
- https://github.com/ronmamo/reflections/issues/373

### CODE REVIEW

**Comment by langchain4j** [comment]
Hi @clear2x almost all tests in `langchain4j-spring-boot-starter` module are failing with this change, could you please check?

**Comment by clear2x** [comment]
Sorry, I have already fixed it.

**Comment by clear2x** [comment]
I found that the `dev.langchain4j.service.spring.mode.automatic.withTools.AiServicesAutoConfigIT#should_create_AI_service_with_tool_that_is_package_private_method_in_package_private_class` case fails to execute both before and after the modification.

File: langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiServicesAutoConfig.java [hunk 1]
```diff
@@ -156,8 +154,10 @@ private static Set<Class<?>> findAiServices(ConfigurableListableBeanFactory bean
         String[] applicationBean = beanFactory.getBeanNamesForAnnotation(SpringBootApplication.class);
         BeanDefinition applicationBeanDefinition = beanFactory.getBeanDefinition(applicationBean[0]);
         String basePackage = applicationBeanDefinition.getResolvableType().resolve().getPackage().getName();
-        Reflections reflections = new Reflections(basePackage);
-        return reflections.getTypesAnnotatedWith(AiService.class);
+        Reflections reflections = new Reflections((new ConfigurationBuilder()).forPackage(basePackage));
+        Set<Class<?>> classes = reflections.getTypesAnnotatedWith(AiService.class);
+        classes.removeIf(clazz -> !clazz.getName().startsWith(basePackage));
```
[the remark this answers is not in the document]
Comment (clear2x) [reply langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiServicesAutoConfig.java:159]:
```java
forPackage(basePackage)); 
```
This code calls the following piece of code : 
```java
public static Collection<URL> forResource(String resourceName, ClassLoader... classLoaders) {
        final List<URL> result = new ArrayList<>();
        final ClassLoader[] loaders = classLoaders(classLoaders);
        for (ClassLoader classLoader : loaders) {
            try {
                final Enumeration<URL> urls = classLoader.getResources(resourceName);
                while (urls.hasMoreElements()) {
                    final URL url = urls.nextElement();
                    int index = url.toExternalForm().lastIndexOf(resourceName);
                    if (index != -1) {
                    	// Add old url as contextUrl to support exotic url handlers
                        result.add(new URL(url, url.toExternalForm().substring(0, index)));
                    } else {
                        result.add(url);
                    }
                }
            } catch (IOException e) {
                if (Reflections.log != null) {
                    Reflections.log.error("error getting resources for " + resourceName, e);
                }
            }
        }
        return distinctUrls(result);
    }
```
where the key part is:
```
int index = url.toExternalForm().lastIndexOf(resourceName);
if (index != -1) {
 // Add old url as contextUrl to support exotic url handlers
    result.add(new URL(url, url.toExternalForm().substring(0, index)));
} else {
    result.add(url);
}
```
This segment extracts a part of the URL, and after extraction, it becomes test-classes.  

Our goal is to find all AI Services under the base package and its sub-packages, so we perform filtering here .

**Comment by clear2x** [comment]
A few other thoughts...... "In jar files, Reflections cannot scan classes." This issue has been discussed extensively in the link above. We might also consider not relying on Reflections and directly using Spring's related methods to achieve our goal.

**Comment by clear2x** [comment]
ResourcePatternResolver, PathMatchingResourcePatternResolver, StandardEnvironment, and SimpleMetadataReaderFactory can work together to achieve our goal.
```java

public class AnnotationUtils extends org.springframework.core.annotation.AnnotationUtils {
    private static final PathMatchingResourcePatternResolver RESOLVER = new PathMatchingResourcePatternResolver();
    private static final SimpleMetadataReaderFactory REGISTER = new SimpleMetadataReaderFactory();
    private static final StandardEnvironment ENVIRONMENT = new StandardEnvironment();

    public static String getResourcePath(String packagePath) {
        if (StringUtils.isEmpty(packagePath)) {
            return "";
        }
        return ResourcePatternResolver.CLASSPATH_ALL_URL_PREFIX
                + ClassUtils.convertClassNameToResourcePath(ENVIRONMENT.resolveRequiredPlaceholders(packagePath))
                + '/' + "**/*.class";
    }

    public static Set<Class<?>> getClazzFromAnnotation(String pkgPath, Class<? extends Annotation> annoClazz) {
        String pathPackage = getResourcePath(pkgPath);
        log.info("pathPackage: {}", pathPackage);
        Set<Class<?>> paths = new HashSet<>();
        Resource[] resources;
        try {
            resources = RESOLVER.getResources(pathPackage);
        } catch (IOException e) {
            return new HashSet<>();
        }
        for (Resource resource : resources) {
            MetadataReader metadataReader = null;
            try {
                metadataReader = REGISTER.getMetadataReader(resource);
            } catch (IOException e) {
                continue;
            }
            AnnotationMetadata annotationMetadata = metadataReader.getAnnotationMetadata();
            if (!annotationMetadata.hasAnnotation(annoClazz.getName())) {
                continue;
            }
            ClassMetadata classMetadata = metadataReader.getClassMetadata();
            String className = classMetadata.getClassName();
            try {
                ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
                Class<?> clazz = classLoader.loadClass(className);
                paths.add(clazz);
            } catch (ClassNotFoundException e) {
                log.warn(e.getMessage());
            }
        }
        return paths;
    }

}


```

**Review by langchain4j — APPROVED** [review]
@clear2x thanks a lot!

**Pull request merged**